### PR TITLE
Additional file descriptor safety checks with refactor of my prior work (issue #5299)

### PR DESCRIPTION
**Pull request by matteius**
## Description
This intends to address problems reported here: #5299
I think these code changes reduce some repeated blocks and leverage the same pattern I added here:  https://github.com/celery/celery/pull/5499/ 

I noticed that the two definitions of on_poll_start had a lot of duplication and so additionally I sought to represent them as a single method that accounts for their one logical difference.

### CODE REVIEW

**Comment by auvipy** [comment]
pydocstyle run-test: commands[0] | pydocstyle /home/travis/build/celery/celery/celery
[7165] /home/travis/build/celery/celery$ /home/travis/build/celery/celery/.tox/pydocstyle/bin/pydocstyle celery
celery/concurrency/asynpool.py:475 in public method `iterate_file_descriptors_safely`:
        D205: 1 blank line required between summary line and description (found 0)
celery/concurrency/asynpool.py:475 in public method `iterate_file_descriptors_safely`:
        D210: No whitespaces allowed surrounding docstring text
celery/concurrency/asynpool.py:475 in public method `iterate_file_descriptors_safely`:
        D400: First line should end with a period (not 's')

File: celery/concurrency/asynpool.py [hunk 1]
```diff
@@ -464,14 +465,56 @@ def _track_child_process(self, proc, hub):
...
+                hub_method(fd, *hub_args, **hub_kwargs)
+            except (OSError, FileNotFoundError) as e:
+                logger.warning(
+                    "Encountered OSError when accessing fd %s ",
+                    fd, exc_info=True)
+                stale_fds.append(fd)  # take note of stale fd
+        # Remove now defunct fds from the managed list
+        if managed_list:
+            for fd in stale_fds:
+                managed_list.pop(fd, None)
```
Comment (tuky) [inline celery/concurrency/asynpool.py]: i am currently trying this out. it did not solve my problem OOB, but i am trying if it works, when the `epoll.unregister` is not swallowing all kinds of exceptions. `managed_list` is a `set`, when being called below with `all_inqueues`, which results in a `TypeError`.

Comment (tuky) [reply celery/concurrency/asynpool.py]: nvm, you already covered this.

Comment (matteius) [reply celery/concurrency/asynpool.py]: Yeah sorry about that -- sent you the link before I added any tests and the existing tests really didn't cover AsyncPool much.

**Comment by matteius** [comment]
@auvipy For some reason I don't know about, the thing this keeps failing on in Travis CI is one of the integration builds because it can't sudo apt-get install some dependency.
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
The command "sudo apt install libcurl4-openssl-dev libssl-dev gnutls-dev" failed and exited with 100 during .

Outside of my pay grade.

**Comment by georgepsarakis** [comment]
@matteius I restarted the failing job.

**Comment by matteius** [comment]
@georgepsarakis  Can you kick off the build again?  I don't have permission to do so, and am forced to make a benign commit to make a new build.

Not sure why it never kicked off the tox component -- it passed the tests that round and seemed to run much longer without doing the tox -- I think Travis CI is having issues this morning because it had some other apt-get install errors outside the scope of these change it did not encounter last night.

File: celery/concurrency/asynpool.py [hunk 2]
```diff
@@ -205,6 +205,59 @@ def _select(readers=None, writers=None, err=None, timeout=0,
             raise
 
 
+try:  # Python 2 does not have FileNotFoundError
```
Comment (thedrow) [inline celery/concurrency/asynpool.py]: Would you mind adding a TODO to drop this once we drop support for Python 2.7?

Comment (matteius) [reply celery/concurrency/asynpool.py]: Done!

**Comment by thedrow** [comment]
Please fix the lint failure.
https://travis-ci.org/celery/celery/jobs/546110436#L549

**Comment by matteius** [comment]
@thedrow  That build was stale, saw that and fixed, please use this one:  https://travis-ci.org/celery/celery/builds/546210878

I think at this point I'll defer and be "Done" making changes, hopefully @tuky  can verify the latest build against the issue reported.

**Comment by matteius** [comment]
@auvipy @thedrow  I just added some more test coverage and was going to try and cover the on_stop_not_started method, but then I searched the code and found that this method is not called from anywhere in the celery project and was written in 2013.   So I deleted it locally and ran all the tests and they passed.   Should that be a separate PR to remove that method or prove that its actually needed?

**Comment by thedrow** [comment]
> @auvipy @thedrow I just added some more test coverage and was going to try and cover the on_stop_not_started method, but then I searched the code and found that this method is not called from anywhere in the celery project and was written in 2013. So I deleted it locally and ran all the tests and they passed. Should that be a separate PR to remove that method or prove that its actually needed?

That method is needed as it is being called by billiard.
See https://github.com/celery/billiard/blob/0e4e7f58618e703e56f1fad2f73076dc1af20ce6/billiard/pool.py#L526.

**Comment by matteius** [comment]
@thedrow Thanks!   We should add some test coverage of it at some point then :-)

**Comment by auvipy** [comment]
can you rebase and add some integration tests for this?

**Comment by matteius** [comment]
@auvipy  So I ran git rebase locally and spent like 20 minutes on it and ended up causing a bunch of conflcits and deleting my repository locally -- what is the point of this exercise and what should I be accomplishing by trying to do it?

**Comment by auvipy** [comment]
that much conflict? !!! to trget of his exercise is to keep the pr codes up to date with master.

**Comment by matteius** [comment]
@auvipy  If the only reason is to keep it up to date with master then wouldn't you just merge master into your branch and no need to rebase?  That is what I did prior here, I have merged the master celery branch into my fork keeping it up to date so there are no merge conflicts.  To me the only reason to rebase is to sqaush commits down -- it might just be easier for me to diff these changes into a new fork off master and make a single commit for you if thats what you want, because rebase just isn't my friend.

**Comment by auvipy** [comment]
I think I meant to say pull from the master but rebase s not needed actually. sorry for misguiding. If you can add an integration test?

**Comment by matteius** [comment]
Yeah, I'll try to get more into the integration tests there to see what can be done.

**Comment by georgepsarakis** [comment]
@matteius `rebase` is preferred so as not to:
- introduce unnecessary merge commits in a PR branch
- maintain a more linear history, which will help future readers

For more details on the differences between merge and rebase you can also read: https://hackernoon.com/git-merge-vs-rebase-whats-the-diff-76413c117333

@auvipy if I may ask, why has the `Create a merge commit` option been disabled? I think it is much easier to track Pull Request merges in the history, and having this information can also be used to extract the changelog.

**Comment by auvipy** [comment]
It was disabled by Ask

**Pull request merged**